Load Collectors by FQCN

DanielBadura opened the pull request:
This patch enables the registry to load collectors by FQCN if not already present in the cache the missing object gets initiated and added to the cache.

Should fix #164

smoench commented:
I had this implementation also in mind then I created the issue. It will works when you directly require deptrac via composer (which is not recommended) but we need to check if it is working with the phar file.

DanielBadura commented:
I will look into it :)

DanielBadura commented on src/Collector/Registry.php:
```diff
@@ -24,11 +24,24 @@ public function __construct($collectors)
+            return $collector;
+        }
+
+        throw new \InvalidArgumentException(sprintf('unknown collector type "%s", possible collectors are %s', $type, implode(', ', array_keys($this->collectors))));
```
should we also update the error message?

smoench replied:
The error message is fine

DanielBadura commented:
I just tested it with a generated phar. I used for this an example depfile (`ControllerServiceRepository1.depfile.yml`) and it worked with it i got the same result. I just changed the type from `className` to the FQCN `SensioLabs\Deptrac\Collector\ClassNameCollector`.

Should i check more?

smoench commented:
Can you create your own / custom collector and use the FQCN in the depfile? I think it doesn't work. Probably we need to check if there is a autoload file and load it. Probably we also need to enable the https://github.com/humbug/php-scoper in Box config.

DanielBadura commented:
Yeah, this is not working. I will look into it later

DanielBadura commented:
I updated the PR to load the autoload file if exists. Maybe you can already have a look on this.
I already added the scoper local but getting some errors. Im going to try to fix it the next days :)

DanielBadura commented:
@smoench i activated the php-scoper with the default configuration and tried to execute deptrac. I get an error which i could not resolve, maybe you can help with that.

```
$ ./deptrac.phar
 0/3 [>---------------------------]   0%PHP Fatal error:  Uncaught TypeError: Return value of _HumbugBoxeae4e57508d1\SensioLabs\Deptrac\AstRunner\AstParser\AstFileReferenceFileCache::get() must be an instance of _HumbugBoxeae4e57508d1\SensioLabs\Deptrac\AstRunner\AstMap\AstFileReference or null, instance of __PHP_Incomplete_Class returned in phar:///srv/share/deptrac.phar/src/AstRunner/AstParser/AstFileReferenceFileCache.php:47
Stack trace:
#0 phar:///srv/share/deptrac.phar/src/AstRunner/AstParser/NikicPhpParser/NikicPhpParser.php(55): _HumbugBoxeae4e57508d1\SensioLabs\Deptrac\AstRunner\AstParser\AstFileReferenceFileCache->get()
#1 phar:///srv/share/deptrac.phar/src/AstRunner/AstRunner.php(30): _HumbugBoxeae4e57508d1\SensioLabs\Deptrac\AstRunner\AstParser\NikicPhpParser\NikicPhpParser->parse()
#2 phar:///srv/share/deptrac.phar/src/Analyser.php(27): _HumbugBoxeae4e57508d1\SensioLabs\Deptrac\AstRunner\AstRunner->createAstMapByFiles()
#3 phar:///srv/share/deptrac.phar/src/Console/Command/AnalyzeCommand.php(54): _HumbugBoxeae4e57508d1\SensioLabs\Deptrac\Analyser->analys in phar:///srv/share/deptrac.phar/src/AstRunner/AstParser/AstFileReferenceFileCache.php on line 47
```

I did some research but found nothing that helped me.

smoench commented:
I think this is a problem with the cache file. With `--no-cache` it should work. In your cache file there might be still the unscoped classnames written but the `allowed_classes` for `unserialize` have the scoped ones -> https://github.com/sensiolabs-de/deptrac/blob/master/src/AstRunner/AstParser/AstFileReferenceFileCache.php#L111

Can you try to add a config for php-scoper with excluding `SensioLabs\Deptrac\*` from scoping -> https://github.com/humbug/php-scoper#namespaces-whitelisting

DanielBadura commented:
You were right, with `--no-cache` i got rid of the error and with the whitelist it is now working 🚀 

So this would be finished from my side and would also fix #19 or not?

smoench commented:
Thank you @DanielBadura! I will try it out on Friday and give you feedback

DanielBadura commented:
@smoench did you had time to try it out?

smoench commented:
Sorry, I was on vacation. I will try it out asap.

DanielBadura commented:
No problem! Hope you had a nice vacation! :)

DanielBadura commented:
Nice! Yeah i will do it the next days :)

DanielBadura commented:
Readme is updated 👍

smoench commented on README.md:
```diff
@@ -596,9 +596,11 @@ layers:
+### Custom Collectors
 
-As deptrac is in a very early state, feel free to contribute your own collector.
+You can even create custom collectors in your project by implementing the `CollectorInterface`.
```
Could you use the FQCN for `CollectorInferface`?

DanielBadura replied:
done https://github.com/sensiolabs-de/deptrac/pull/320/commits/617a1fd835ffcce455c521c5ef7023fdd659e572

smoench commented:
Thank you @DanielBadura !
